### PR TITLE
fix high tag encoding, bad input handling

### PR DESCRIPTION
**Pull request by judwhite**

### CODE REVIEW

File: identifier.go [hunk 1]
```diff
@@ -80,24 +79,24 @@ func encodeIdentifier(identifier Identifier) []byte {
...
 		tag := identifier.Tag
 
-		highBit := uint(63)
-		for {
-			if tag&(1<<highBit) != 0 {
-				break
-			}
-			highBit--
-		}
+		b = append(b, encodeHighTag(tag)...)
```
Comment (johnweldon) [inline identifier.go]: Nice extracting to a function

File: identifier.go [hunk 2]
```diff
@@ -80,24 +79,24 @@ func encodeIdentifier(identifier Identifier) []byte {
...
-			offset := uint(i) * 7
-			mask := Tag(0x7f) << offset
-			tagByte := (tag & mask) >> offset
-			if i != 0 {
-				tagByte |= 0x80
-			}
-			b = append(b, byte(tagByte))
+func encodeHighTag(tag Tag) []byte {
+	b := make([]byte, 0, 4)
+	for tag != 0 {
```
Comment (johnweldon) [inline identifier.go]: This seems like a cleaner implementation (and similar to other implementations I've seen in python for example)  I'll have to take some time to actually understand it though.  I trust the tests are robust enough to demonstrate that this is at least as good as the old implementation.

File: identifier.go [hunk 3]
```diff
@@ -80,24 +79,24 @@ func encodeIdentifier(identifier Identifier) []byte {
...
+		t := tag & HighTagValueBitmask
+		tag >>= 7
+		if len(b) != 0 {
+			t |= HighTagContinueBitmask
 		}
+		b = append(b, byte(t))
+	}
+	// reverse
```
Comment (johnweldon) [inline identifier.go]: ... and I have no idea why we have to reverse it here, (endianness?) but when I take the time to actually understand this I assume it will make sense.

File: ber.go [hunk 4]
```diff
@@ -9,6 +9,8 @@ import (
 	"reflect"
 )
 
+const maxPacketLengthBytes = 0x8000000 // 134,217,728 bytes (chosen arbitrarily)
```
Comment (johnweldon) [inline ber.go]: Seems reasonable

Comment (liggitt) [reply ber.go]: Ldap search results come back as a single packet, right? I’ve seen some surprisingly large result sets that would definitely exceed this

Comment (judwhite) [reply ber.go]: @liggitt I'm looking for some guidance here. What would be reasonable, 1-2GB? We could also make it a package level variable if a user encounters a use can which needs a higher value.

The issue is `make([]byte, length)` panics if there's not enough memory. The intention here is to protect ourselves from an errant server or network which returns an obviously invalid result. What's "obviously invalid" isn't always so obvious.

References for `make()` and its limits:
- https://golang.org/src/runtime/slice.go#L31
- https://golang.org/src/runtime/malloc.go#L137

File: identifier.go (hunk 2 above)
Comment (judwhite) [reply identifier.go]: @johnweldon This part is the bug fix. I didn't dive deep on the previous implementation but it looked like an OBOE. The [Travis output](https://travis-ci.org/go-asn1-ber/asn1-ber/jobs/279164682#L506-L512) shows it munching away bytes in the unpack-pack-unpack operations.

```
--- FAIL: TestSuiteDecodePacket (0.00s)
	suite_test.go:118: tests/tc49.ber: data should be the same
		want: []byte{0x9f, 0x81, 0x6, 0x0}
		got: []byte{0x9f, 0x6, 0x0}
	suite_test.go:130: tests/tc49.ber: data should be the same
		want: []byte{0x9f, 0x6, 0x0}
		got: []byte{0x86, 0x0}
```

Here's an annotated version of `encodeHighTag`:

```go
func encodeHighTag(tag Tag) []byte {
	// set cap=4 so hopefully we don't have to do another allocation
	b := make([]byte, 0, 4)
	for tag != 0 {
		// t := last 7 bits of tag (HighTagValueBitmask = 0x7F)
		t := tag & HighTagValueBitmask

		// rshift tag 7 to remove what we just pulled off
		tag >>= 7

		// if we previously added to b this entry needs a continuation bit (0x80)
		if len(b) != 0 {
			// add the continuation bit
			t |= HighTagContinueBitmask
		}

		b = append(b, byte(t))
	}
	// reverse
	// since we pulled bits off tag small to high our byte slice is in the wrong order
	// example: tag = 0xFF results in {0x7F, 0x01 + 0x80 (continuation bit)}
	// this needs to be reversed into 0x81 0x7F
	for i, j := 0, len(b)-1; i < len(b)/2; i++ {
		b[i], b[j-i] = b[j-i], b[i]
	}
	return b
}
```

File: ber.go (hunk 4 above)
Comment (liggitt) [reply ber.go]: > @liggitt I'm looking for some guidance here. What would be reasonable, 1-2GB? We could also make it a package level variable if a user encounters a use can which needs a higher value.

yeah... it's tricky. there's basically four modes of operating:
* stream results to disk, support partial reads/lazy-loading/on-demand-parsing. requires fundamentally changing the API, not really an option here.
* allow arbitrarily large allocations on a system with paging enabled, allocs larger than available physical memory are (sometimes) tolerated by the underlying OS, other times oom
* allow arbitrarily large allocations on a system without paging enabled, allocs larger than available physical memory oom
* artificially set a max size in the library. this is essentially impossible to do correctly for all environments (in some embedded systems, 100MB could still fail, in other systems, 2GB of search results is perfectly tolerable)

We should also decide what we're protecting against:
* malicious server? if so, they could just set indeterminate length and stream forever anyway, so this doesn't really help.
* spurious data that flips a bit in the length descriptor? if so, we could simply set a max size on the *initial* buffer we accumulate into, but let it grow as we read the data. this would protect against a case where a malformed 1k packet claimed to have a 1TB of data, since we'd EOF and error before actually needing to allocate a buffer big enough to hold that.

I'd be in favor of either of the following:
* max initial buffer size, allow growing while reading packet
* package-level absolute max size, settable by consumers, defaulting to no limit (as today)

Comment (johnweldon) [reply ber.go]: I like the idea of a package variable.

File: identifier.go (hunk 2 above)
Comment (liggitt) [reply identifier.go]: +1 for the annotated version... can you add the comments to the commit?

File: ber.go (hunk 4 above)
Comment (judwhite) [reply ber.go]: > "sane" max initial buffer size (1MB? 10MB?), allow growing while reading packet

@liggitt @johnweldon I'd prefer this also, but it's additional work which needs to be done carefully. I'd recommend doing this in a follow-up. I'll be upfront, I can't promise I'll be the one to do it 😄 

Maybe using [io.LimitedReader](https://golang.org/pkg/io/#LimitedReader) would help.

The motivation was to limit panics which could crash the application, however they may come about. If we start getting into how that may happen it usually derails the conversation (unless we're specifically doing threat modeling).

> package-level absolute max size, settable by consumers, defaulting to no limit (as today)

I think this is a sane quick fix. In the test it attempts to allocated 200+ GB, which I would assume is outside even most edge case uses.

For now I changed it to an exported variable with a default of `math.MaxInt32` (~2GB). It works on my 64-bit machine, and Travis shouldn't have any issues because there isn't a test for allocating the max allowed byte slice.

Please let me know if this is acceptable or if you'd like to try something else.

**Comment by judwhite** [comment]
@johnweldon @liggitt Any feedback? Last comment is collapsed in a "show outdated" https://github.com/go-asn1-ber/asn1-ber/pull/13#discussion_r140668584

File: ber.go (hunk 4 above)
Comment (johnweldon) [reply ber.go]: I think it's acceptable.  Thank you again for your contributions.  I'll make sure @liggitt doesn't have any reservations, and then we'll merge this in.

File: ber.go [hunk 5]
```diff
@@ -5,10 +5,14 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"math"
 	"os"
 	"reflect"
 )
 
+// MaxPacketLengthBytes specifies the maximum allowed packet size when calling ReadPacket or DecodePacket.
+var MaxPacketLengthBytes int = math.MaxInt32
```
Comment (liggitt) [inline ber.go]: make this type int64 for future compatibility

File: ber.go [hunk 6]
```diff
@@ -330,6 +334,9 @@ func readPacket(reader io.Reader) (*Packet, int, error) {
 	}
 
 	// Read definite-length content
+	if length > MaxPacketLengthBytes {
```
Comment (liggitt) [inline ber.go]: `if MaxPacketLengthBytes > 0 && int64(length) > MaxPacketLengthBytes`, and doc the variable as "set to 0 for no limit"

File: ber.go (hunk 5 above)
Comment (judwhite) [reply ber.go]: @liggitt I'll try, not sure how this will work when `GOARCH=386`, may need to convert it.

Somewhat related, in `.travis.yml` you can test other OS's/architectures. I think this might be useful now since we're using `int64` and a builtin. Travis' `env` builds a matrix with all the Go versions, do you have an idea which `GOOS` / `GOARCH` combinations you'd like to target and which Go versions to keep?

```
env:
  - GOOS=linux GOARCH=amd64
  - GOOS=windows GOARCH=amd64
  - GOOS=linux GOARCH=386
  - GOOS=windows GOARCH=386
```

Comment (johnweldon) [reply ber.go]: We've already had a bug in the past with 386 arch; so I think adding that to the test matrix would be a win.

It's appealing to me to keep testing as far back as we do (1.2).

Comment (judwhite) [reply ber.go]: @johnweldon Oops, missed this comment before pushing `.travis.yml`. Let's see how this one goes (I know it'll build Windows binaries but not sure it can run tests) then I'll add back the ones I dropped.

File: suite_test.go [hunk 7]
```diff
@@ -80,8 +81,17 @@ var testcases = []struct {
 	{File: "tests/tc47.ber", Error: "eoc child not allowed with definite length"},
 	{File: "tests/tc48.ber", Error: "", IndefiniteEncoding: true}, // Error: "Using of more than 7 "unused bits" in BIT STRING with constrictive encoding form"
 	{File: "tests/tc49.ber", Error: ""},
-	{File: "tests/tc50.ber", Error: "length cannot be less than -1"},
-	{File: "tests/tc51.ber", Error: fmt.Sprintf("length %d greater than maximum %d", 206966894640, MaxPacketLengthBytes)},
+	{File: "tests/tc50.ber", Error: is64bit("length cannot be less than -1", "long-form length overflow")},
+	{File: "tests/tc51.ber", Error: is64bit(fmt.Sprintf("length 206966894640 greater than maximum %v", MaxPacketLengthBytes), "long-form length overflow")},
```
Comment (judwhite) [inline suite_test.go]: @johnweldon @liggitt Open to suggestions. The error message is different between `GOARCH=386` and `GOARCH=amd64`.

**Comment by judwhite** [comment]
@johnweldon @liggitt When you guys are good I'll squash this down into a reasonable set of commits.

**Comment by liggitt** [comment]
We don’t need to test the full grid of go versions and architectures. I’d probably fully test the endpoints and a single os/arch for the intermediate go versions

**Comment by johnweldon** [comment]
Agreed about not needing to test the full matrix; does Travis support 'cherry-picking' combinations to test?

Also, do we need to figure out how to fix the 386 error right away?

**Comment by liggitt** [comment]
See “exclude” - https://docs.travis-ci.com/user/customizing-the-build#Build-Matrix

**Comment by judwhite** [comment]
> Also, do we need to figure out how to fix the 386 error right away?

@johnweldon The `amd64` and `386` both error on the 200+GB packet, it's just a different error. They could be brought in line if we returned `int64` from `readLength` and did the necessary checks. ¯\\_(ツ)_/¯

> See “exclude” - https://docs.travis-ci.com/user/customizing-the-build#Build-Matrix

@liggitt Updated before I saw this. Would exclude make `.travis.yml` more succinct/clear, or are you happy with what's there?

Let me know when you're ready for a squash.

**Comment by liggitt** [comment]
> Would exclude make .travis.yml more succinct/clear, or are you happy with what's there?

I commented while mobile, I prefer your `include`

**Comment by judwhite** [comment]
@liggitt @johnweldon Thanks for your help working through this. Want me to squash these commits into a more logical history?

**Comment by johnweldon** [comment]
@judwhite yes - thanks.  Let's get this wrapped up and merged

**Comment by judwhite** [comment]
@johnweldon @liggitt History cleaned up, ready for review.

**Comment by liggitt** [comment]
LGTM, @johnweldon has merge

**Comment by johnweldon** [comment]
Thank you again @judwhite 
Thanks @liggitt also!
